# Actividad práctica 

### Instrucciones
- Escriba las rutinas de Python necesarias para resolver los problemas de cada punto
- Siga las instrucciones y conteste donde se pida
- Entregue el notebook con sus respuestas al correo: phuijse@inf.uach.cl
- El trabajo es individual

### Introducción
En esta actividad usaremos la base de datos [*Auto Milles per Gallon Data Set*](https://archive.ics.uci.edu/ml/datasets/auto+mpg) del repositorio UCI

Puede descargarla usando el siguiente bloque

In [1]:
%%bash
wget -O auto.csv -c https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data-original 
wget -O README -c https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.names 

mesg: ttyname failed: Inappropriate ioctl for device
--2019-05-09 00:35:41--  https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data-original
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 32149 (31K) [application/x-httpd-php]
Saving to: ‘auto.csv’

     0K .......... .......... .......... .                    100% 80.1K=0.4s

2019-05-09 00:35:43 (80.1 KB/s) - ‘auto.csv’ saved [32149/32149]

--2019-05-09 00:35:43--  https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.names
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1660 (1.6K) [application/x-httpd-php]
Saving to: ‘README’

     0K .         

### Información preliminar

1. Use las herramientas de UNIX para leer el archivo README en el notebook

In [7]:
%%bash
cat /mnt/c/Users/Pricila/INFO147-master/unidad1/tarea/data/README

1. Title: Auto-Mpg Data

2. Sources:
   (a) Origin:  This dataset was taken from the StatLib library which is
                maintained at Carnegie Mellon University. The dataset was 
                used in the 1983 American Statistical Association Exposition.
   (c) Date: July 7, 1993

3. Past Usage:
    -  See 2b (above)
    -  Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning.
       In Proceedings on the Tenth International Conference of Machine 
       Learning, 236-243, University of Massachusetts, Amherst. Morgan
       Kaufmann.

4. Relevant Information:

   This dataset is a slightly modified version of the dataset provided in
   the StatLib library.  In line with the use by Ross Quinlan (1993) in
   predicting the attribute "mpg", 8 of the original instances were removed 
   because they had unknown values for the "mpg" attribute.  The original 
   dataset is available in the file "auto-mpg.data-original".

   "The data concerns city-cycle fuel consumptio

2. Responda ¿Cuántos y cuáles son los atributos (características)? ¿De qué tipo es cada uno?

Los atributos son 9 y, sus nombres, tipo y descripción respectiva son:    
    
         NOMBRE           TIPO                             DESCRIPCIÓN
     
    1. mpg:           continuous                         Rendimiento ("miles per gallon")  
    2. cylinders:     multi-valued discrete              Número de cilindros
    3. displacement:  continuous                         Desplazamiento
    4. horsepower:    continuous                         Caballos de fuerza
    5. weight:        continuous                         Peso
    6. acceleration:  continuous                         Aceleración                        
    7. model year:    multi-valued discrete              Año de modelo
    8. origin:        multi-valued discrete              Origen ?
    9. car name:      string (unique for each instance)  Nombre

3. Responda ¿Cuantas instancias/ejemplos existen?

### Lectura de datos

1. Cree un `DataFrame` a partir de `auto.csv` 
    1. Use una columna por atributo* 
    1. Use los nombres de columna descritos en `README` 
1. Imprima las primeras 5 filas del dataframe

**HINT:** Verifique el uso  del argumento `delim_whitespace`*

In [428]:
import pandas as pd
arc = pd.read_csv("Data/auto.csv", delim_whitespace=True, names=["mpg","cylinders","displacement",
                                                                "horsepower","weight","acceleration",                                                                
                                                                "year","origin","name"])
df = pd.DataFrame(arc)
df[:5]

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0        8.0         307.0       130.0  3504.0          12.0  70.0   
1  15.0        8.0         350.0       165.0  3693.0          11.5  70.0   
2  18.0        8.0         318.0       150.0  3436.0          11.0  70.0   
3  16.0        8.0         304.0       150.0  3433.0          12.0  70.0   
4  17.0        8.0         302.0       140.0  3449.0          10.5  70.0   

   origin                       name  
0     1.0  chevrolet chevelle malibu  
1     1.0          buick skylark 320  
2     1.0         plymouth satellite  
3     1.0              amc rebel sst  
4     1.0                ford torino

### Manipulación de columnas 


**1. Convierta las columnas *cylinders*, *year* y *origin* a valores enteros**


In [429]:
df.dtypes

mpg             float64
cylinders       float64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
year            float64
origin          float64
name             object
dtype: object

In [430]:
#conv = dict.fromkeys(['cylinders', 'year', 'origin'], lambda x: int(x.strip(" ")))

df["year"] = df["year"].astype("int")
df["origin"] = df["origin"].astype("int")
df["cylinders"] = df["cylinders"].astype("int")

df.dtypes

mpg             float64
cylinders         int32
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
year              int32
origin            int32
name             object
dtype: object

**2. Reemplace los valores $[1, 2, 3]$ de *origin* por *USA*, *Europe*, *Japan*, respectivamente**

In [431]:
df["origin"].replace({1: "USA", 2: "Europe", 3: "Japan"},inplace=True)
df[:5]

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0  3504.0          12.0    70   
1  15.0          8         350.0       165.0  3693.0          11.5    70   
2  18.0          8         318.0       150.0  3436.0          11.0    70   
3  16.0          8         304.0       150.0  3433.0          12.0    70   
4  17.0          8         302.0       140.0  3449.0          10.5    70   

  origin                       name  
0    USA  chevrolet chevelle malibu  
1    USA          buick skylark 320  
2    USA         plymouth satellite  
3    USA              amc rebel sst  
4    USA                ford torino

**3. Convierta los valores de la columna *mpg* (galones por milla) a kmpl (kilómetros por litro)**


In [432]:
def conv_kpl(mpg):
    '''
    Para convertir mpg a kpl, multiplique sus millas por galón (mpg) con 0.4251 
    18.0 (mpg) -> 7.6518 (kpl)
    ''' 
    kml = mpg*0.4251
#    print(kml)
    return kml

df["mpg"]= df["mpg"].apply(lambda x: conv_kpl(x))
df["mpg"].head()

0    7.6518
1    6.3765
2    7.6518
3    6.8016
4    7.2267
Name: mpg, dtype: float64

**4. Convierta los valores de la columna peso de libras (*pounds*) a kilogramos**


In [433]:
def conv_kg(weight):
    ''' Para convertir lb a kg, divida las libras (lb) en 2.046
        3504 lb -> 1589.388 kg
    '''
    kg = weight/2.2046
    return kg

df["weight"] = df["weight"].apply(lambda x: conv_kg(x))

df["weight"].head()

0    1589.403974
1    1675.133811
2    1558.559376
3    1557.198585
4    1564.456137
Name: weight, dtype: float64

**5. Modifique el nombre de la columna *mpg* a kmpl**


In [434]:
df = df.rename(columns = {'mpg':'kpl'})
df["kpl"].head()

0    7.6518
1    6.3765
2    7.6518
3    6.8016
4    7.2267
Name: kpl, dtype: float64

### Datos perdidos



**1. Encuentre los nombres de los autos que tengan valores perdidos (*missing data*)**


In [486]:
df[df.isnull().any(axis=1)].name

AttributeError: 'DataFrame' object has no attribute 'name'

**2. Encuentre los nombres de los atributos que tengan valores perdidos**


In [436]:
df.isnull().sum()

kpl             8
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
year            0
origin          0
name            0
dtype: int64

**kpl** y **horsepower** tienen valores perdidos con 8 y 6 valores respectivamente

**3. Elimines las filas con atributos perdidos ¿Cuántas filas quedan en el dataframe?**

In [437]:
df.dropna(inplace=True)
df.shape

(392, 9)

*392* filas y 9 columnas, antes 406 filas

### Multi-índice

**HINT:** Verifique el uso de `.str.split()`* y `str.replace()`**

**1. Cree un dataframe con dos columnas a partir del atributo "*car name*"**
    1. La primera columna debe tener la primera palabra del string en car name* 
    1. La segunda columna debe tener el resto del string en car name

In [438]:
df1 = pd.DataFrame()
df1[0] = df["name"].str.split(n=1, expand = True)[0]
df1[1] = df["name"].str.split(n=1, expand = True)[1]

df1[:5]

0                1
0  chevrolet  chevelle malibu
1      buick      skylark 320
2   plymouth        satellite
3        amc        rebel sst
4       ford           torino

**2. Elimine la columna *car name* del dataframe y agregue estas nuevas columnas llamándolas *brand* y *model*, respectivamente**

In [439]:
df.drop(['name'], axis=1, inplace=True)
df = pd.concat([df,df1],axis=1)
df = df.rename(columns = {0:'brand',1:'model'})
df[:5]

kpl  cylinders  displacement  horsepower       weight  acceleration  \
0  7.6518          8         307.0       130.0  1589.403974          12.0   
1  6.3765          8         350.0       165.0  1675.133811          11.5   
2  7.6518          8         318.0       150.0  1558.559376          11.0   
3  6.8016          8         304.0       150.0  1557.198585          12.0   
4  7.2267          8         302.0       140.0  1564.456137          10.5   

   year origin      brand            model  
0    70    USA  chevrolet  chevelle malibu  
1    70    USA      buick      skylark 320  
2    70    USA   plymouth        satellite  
3    70    USA        amc        rebel sst  
4    70    USA       ford           torino

**3. ¿Cuántas marcas (*brand*) únicas existan? ¿Existen columnas repetidas o mal escritas? Mencione cuales y corrijalas**

In [440]:
#df['brand'].duplicated()
#df.groupby(df.brand,as_index=False).size()
#df.duplicated(subset='brand', keep='first').sum()
df.groupby('brand').size().loc[lambda x: x==1]

brand
capri        1
chevroelt    1
hi           1
mercedes     1
nissan       1
toyouta      1
triumph      1
vokswagen    1
dtype: int64

In [441]:
df.groupby(df.brand).size()

brand
amc              27
audi              7
bmw               2
buick            17
cadillac          2
capri             1
chevroelt         1
chevrolet        43
chevy             3
chrysler          6
datsun           23
dodge            28
fiat              8
ford             48
hi                1
honda            13
maxda             2
mazda            10
mercedes          1
mercedes-benz     2
mercury          11
nissan            1
oldsmobile       10
opel              4
peugeot           8
plymouth         31
pontiac          16
renault           3
saab              4
subaru            4
toyota           25
toyouta           1
triumph           1
vokswagen         1
volkswagen       15
volvo             6
vw                6
dtype: int64

Errores:

        1. chevroelt  ->  chevrolet
        2. maxda      ->  mazda
        3. mercedes   ->  mercedes-benz
        4. toyouta    ->  toyota
        5. vokswagen  ->  volkswagen

In [442]:
df.replace({"chevroelt":"chevrolet","maxda":"mazda","mercedes":"mercedes-benz",
            "toyouta":"toyota","vokswagen":"volkswagen"}, inplace=True)
df.groupby(df.brand).size()

brand
amc              27
audi              7
bmw               2
buick            17
cadillac          2
capri             1
chevrolet        44
chevy             3
chrysler          6
datsun           23
dodge            28
fiat              8
ford             48
hi                1
honda            13
mazda            12
mercedes-benz     3
mercury          11
nissan            1
oldsmobile       10
opel              4
peugeot           8
plymouth         31
pontiac          16
renault           3
saab              4
subaru            4
toyota           26
triumph           1
volkswagen       16
volvo             6
vw                6
dtype: int64

**4. Cree un multi-índice que utilice como primera llave a *brand* y como segunda a *model***

In [459]:
df.set_index(["brand","model"],inplace=True)

**5. Muestre los autos con marca *Mazda***

In [477]:
df.loc["mazda"]

kpl  cylinders  displacement  horsepower       weight  \
model                                                                      
rx2 coupe      8.07690          3          70.0        97.0  1056.881067   
rx3            7.65180          3          70.0        90.0   963.440080   
rx-4           9.13965          3          80.0       110.0  1233.783906   
glc deluxe    13.94328          4          78.0        52.0   900.390093   
glc deluxe    14.49591          4          86.0        65.0   895.854123   
626           13.30563          4         120.0        75.0  1153.043636   
glc           19.80966          4          86.0        65.0   957.089721   
rx-7 gs       10.07487          3          70.0       100.0  1097.704799   
glc 4         14.49591          4          91.0        68.0   900.390093   
626           13.43316          4         120.0        74.0  1195.228159   
glc custom l  15.72870          4          91.0        68.0   918.533974   
glc custom    13.17810          4          91.0        68.0   893.586138   

              acceleration  year origin  
model                                    
rx2 coupe             13.5    72  Japan  
rx3                   13.5    73  Japan  
rx-4                  13.5    77  Japan  
glc deluxe            19.4    78  Japan  
glc deluxe            15.2    79  Japan  
626                   17.5    80  Japan  
glc                   17.9    80  Japan  
rx-7 gs               12.5    80  Japan  
glc 4                 16.0    81  Japan  
626                   18.3    81  Japan  
glc custom l          18.2    82  Japan  
glc custom            17.6    82  Japan

**6. Usando groupby cuente la cantidad de modelos por marca**

In [502]:
df.groupby(['brand','model']).size()

brand       model                 
amc         ambassador brougham       1
            ambassador dpl            1
            ambassador sst            1
            concord                   2
            concord d/l               1
            concord dl 6              1
            gremlin                   4
            hornet                    4
            hornet sportabout (sw)    1
            matador                   5
            matador (sw)              2
            pacer                     1
            pacer d/l                 1
            rebel sst                 1
            spirit dl                 1
audi        100 ls                    1
            100ls                     2
            4000                      1
            5000                      1
            5000s (diesel)            1
            fox                       1
bmw         2002                      1
            320i                      1
buick       century                   2
     

### Buscando autos

1. Por cada marca encuentre el auto más pesado y el más liviano
1. Por cada marca encuentre los autos con mayor y menor caballos de fuerza, respectivamente
1. Por cada marca encuentre los autos con menor y mayor desempeño en términos de kmpl, respectivamente

**1. Por cada marca encuentre el auto más antiguo y el más nuevo**

In [536]:
grupo = df.groupby("brand")
grupo["year"].max()

brand
amc              80
audi             80
bmw              77
buick            82
cadillac         79
capri            76
chevrolet        82
chevy            82
chrysler         82
datsun           82
dodge            82
fiat             79
ford             82
hi               70
honda            82
mazda            82
mercedes-benz    80
mercury          82
nissan           82
oldsmobile       82
opel             76
peugeot          81
plymouth         82
pontiac          82
renault          77
saab             78
subaru           81
toyota           82
triumph          80
volkswagen       82
volvo            81
vw               82
Name: year, dtype: int32

In [537]:
grupo["year"].min()

brand
amc              70
audi             70
bmw              70
buick            70
cadillac         76
capri            76
chevrolet        70
chevy            70
chrysler         72
datsun           70
dodge            70
fiat             71
ford             70
hi               70
honda            74
mazda            72
mercedes-benz    76
mercury          71
nissan           82
oldsmobile       72
opel             71
peugeot          70
plymouth         70
pontiac          70
renault          72
saab             70
subaru           74
toyota           70
triumph          80
volkswagen       70
volvo            72
vw               76
Name: year, dtype: int32

### Análisis exploratorio

1. Encuentre el promedio y la desviación estándar de cada atributo numérico en toda la tabla y por marca
1. Construya gráficos de barras del rendimiento (kmpl) promedio en función del año, peso, origen,  número de cilindros y marca (use *barras de error* con la desviación estándar del rendimiento)
1. Viendo estos gráficos, ¿Qué puede concluir respecto a la relación del rendimiento con estos factores?

### Gráficos interactivos

1. Construya una interfaz gráfica usando `ipywidgets` que permita seleccionar dos atributos y los gráfique como un *scatter plot*
    1. El primer atributo será el eje de abscisas y el segundo atributo el eje de ordenadas
    1. Ese el atributo de origen como el color de los marcadores
1. Los gráficos deben ser legibles, los ejes deben estar etiquetados, use leyenda para los marcadores